signac/contrib/project.py

```diff
@@ -95,6 +95,22 @@ def find_job_ids(self, filter=None, doc_filter=None):
         return self._collection._find(filter)
 
 
+class ProjectConfig(Config):
+    """Extends the project config to make it immutable."""
+    def __init__(self, *args, **kwargs):
+        self._mutable = True
+        super(ProjectConfig, self).__init__(*args, **kwargs)
+        self._mutable = False
+
+    def __setitem__(self, key, value):
+        if not self._mutable:
+            warnings.warn("Modifying the project configuration after project "
+                          "initialization is deprecated as of version 1.3 and "
+                          "will be removed in version 2.0.",
+                          DeprecationWarning)
```

Comment:
We could add a unit tests and the following assertion, so that we don't forget to actually remove this with 2.0:
```suggestion
                          DeprecationWarning)
    from packaging import version
    assert  version.parse(__version__) < version.parse("2.0")
```

Reply:
What would you want the unit test for? Adding both a unit test and the assertion seems redundant to me, unless you mean adding a unit test that for now checks for a warning and later does an `assertRaises`. Since removing config modification will require modifying a few tests anyway, I'd prefer to just hold off on adding that test until 2.0 when we update all the tests for this change.

Reply:
If we add the assertion we need to make sure that this function is actually executed in *any* of the tests. Otherwise we risk that we miss this and our code starts to fail with an `AssertionException`. The point is that the test fails with 2.0 and we remember to remove this.

A natural test would be to simply access and optionally attempt to modify the configuration.